pkg/apis/storage/validation/validation.go

```diff
@@ -441,3 +442,21 @@ func validatePodInfoOnMount(podInfoOnMount *bool, fldPath *field.Path) field.Err
 
 	return allErrs
 }
+
+// validateVolumeLifecycleModes tests if mode has one of the allowed values.
+func validateVolumeLifecycleModes(modes []storage.VolumeLifecycleMode, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
```

Comment:
Validation needs to be gated on feature flag enablement.  If the flag is off, this validation should be a no-op.

Reply:
When doing that, this (currently) invalid test case no longer triggers an error:
```
		{
			// invalid mode
			ObjectMeta: metav1.ObjectMeta{Name: driverName},
			Spec: storage.CSIDriverSpec{
				AttachRequired: &attachNotRequired,
				PodInfoOnMount: &notPodInfoOnMount,
				VolumeLifecycleModes: []storage.VolumeLifecycleMode{
					"no-such-mode",
				},
			},
		},
```

Is that really intended? Just wondering, I already updated `TestCSIDriverValidation` accordingly and now run it with and without the feature enabled, to cover both scenarios.

See commit a5bd16d9fba9686e0e49ec53bf7779dd83adc7cf.

I'll change `TestCSIDriverValidationUpdate` in the upcoming testing PR. In the meantime, can we merge this PR as it is now? I'm currently on vacation and would prefer to get this merged with just those changes that are considered essential.

It also continues to conflict with other changes going into master (like 7c1f55f4003ae24ffbdbb8ca6f6594807783230a). I rebased to resolved that conflict. @msau42, can you re-add your lgtm?

Reply:
Wait, this also affect `TestCSIDriverValidation/invalid_volume_mode` in `strategy_test.go`. I'll address that now.

Reply:
Done.

Reply:
The scenario is that someone can turn the gate on, get a value stored, then turn the gate off.  If the gate is off, the object from storage must be considered valid even if the field is set, or set incorrectly.

Reply:
> Validation needs to be gated on feature flag enablement. If the flag is off, this validation should be a no-op.

there are a couple existing exceptions in the codebase, but in general, validation should not pay attention to feature gates.

The recommended approach:
1. In PrepareForCreate data related to a disabled feature
2. In PrepareForUpdate data related to a disabled feature *unless* the existing object already made use of that data
3. In validation, validate normally without regard to feature gates

Reply:
see https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#alpha-field-in-existing-api-version for details

Reply:
Thanks for the clarification Jordan

Reply:
Please remove the gate as Jordan describes.

Reply:
You may be able to drop this entire commit (if there were no new tests added)

Reply:
I removed the commit.

@smarterclayton, @msau42 : please check again and lgtm if it's ready.